client/js/geometry.js

```diff
@@ -18,12 +18,58 @@ export function getOffset(origin, target) {
   return {x: target.x - origin.x, y: target.y - origin.y}
 }
 
-export function applyTransformedOffset(origin, offset, scale, rotation) {
-  if(rotation % 360 == 0) {
-    return {x: origin.x + offset.x * scale, y: origin.y + offset.y * scale}
-  } else {
-    const dist = distance({x:0,y:0},offset) * scale;
-    const angle = Math.atan2(offset.y ,offset.x) + rotation * Math.PI / 180;
-    return {x: origin.x + dist * Math.cos(angle), y: origin.y + dist * Math.sin(angle)};
+export function getScreenTransform(elem) {
+  let transform = new DOMMatrix();
+  let t = new DOMPoint(10, 10);
+  while (elem) {
+    transform.preMultiplySelf(getElementTransform(elem));
+    elem = elem.offsetParent;
   }
+  return transform;
+}
+
+function parseLengths(str) {
+  return str.split(' ').map(s => parseFloat(s))
+}
+
+export function dehomogenize(point) {
+  return new DOMPoint(point.x / point.w, point.y / point.w, point.z / point.w);
+}
+
+export function getPointOnPlane(transform, x, y) {
+  const inv = transform.inverse();
+  const p0 = dehomogenize(inv.transformPoint(new DOMPoint(x, y, 0)));
+  // Get transformed direction vector of ray in (0, 0, 1) direction.
+  let dir = dehomogenize(inv.transformPoint(new DOMPoint(x, y, 1)));
+  dir.x -= p0.x;
+  dir.y -= p0.y;
+  dir.z -= p0.z;
+  // If the projected line is parallel with the plane, no intersection point can be determined.
```

Comment:
Note we generally shouldn't hit this case since it is invalid (and for dragging you shouldn't be able to hit test to a plane / widget that is completely showing its side to you as it is not visible), but we may want to do something to more gracefully handle the null return so it doesn't result in a JS error.

Reply:
Done in latest commit.